### PR TITLE
Program arg support for run configs (799).

### PR DESCRIPTION
**Pull request by pq**
Fixes: #799.

@skybrian @devoncarew

### CODE REVIEW

File: src/io/flutter/run/SdkFields.java [hunk 1]
```diff
@@ -25,13 +25,15 @@
  */
 public class SdkFields {
   private @Nullable String filePath;
+  private @Nullable String programArgs;
   private @Nullable String workDir;
 
-  public SdkFields() {}
+  public SdkFields() {
```
Comment (pq) [inline src/io/flutter/run/SdkFields.java]: @skybrian : sorry for the unrelated formatting changes -- I ran the formatter (which we generally do) and this happened.  Hopefully not too distracting...

File: src/io/flutter/run/SdkFields.java [hunk 2]
```diff
@@ -59,7 +70,7 @@ public void setWorkingDirectory(final @Nullable String dir) {
 
   /**
    * Reports any errors that the user should correct.
-   *
+   * <p>
```
Comment (skybrian) [inline src/io/flutter/run/SdkFields.java]: Also here.

File: src/io/flutter/run/SdkFields.java [hunk 3]
```diff
@@ -25,13 +25,15 @@
...
 
-  public SdkFields() {}
+  public SdkFields() {
+  }
 
   /**
    * Creates SDK fields from a Dart file containing a main method.
-   *
+   * <p>
    * <p>(Automatically chooses the working directory.)
```
Comment (skybrian) [inline src/io/flutter/run/SdkFields.java]: Remove redundant `<p>`

File: src/io/flutter/run/SdkFields.java [hunk 4]
```diff
@@ -77,7 +88,8 @@ VirtualFile chooseWorkDir(VirtualFile launchFile, Project project) throws Runtim
     // Work directory is optional; if not set we must infer it.
     if (!StringUtil.isEmptyOrSpaces(workDir)) {
       return checkWorkDir(workDir);
-    } else {
+    }
+    else {
```
Comment (skybrian) [inline src/io/flutter/run/SdkFields.java]: Okay but I wonder if we could tweak formatter rules to use `} else {`?

This is Google style so I keep forgetting to do it the Jetbrains way.

File: src/io/flutter/run/daemon/FlutterAppService.java [hunk 5]
```diff
@@ -45,17 +44,19 @@ private FlutterAppService(Project project) {
...
    */
   public FlutterApp startFlutterSdkApp(@NotNull String workDir,
+                                       @Nullable String programArgs,
                                        @Nullable FlutterDevice device,
                                        @NotNull RunMode mode,
                                        @Nullable String path)
     throws ExecutionException {
 
-    final GeneralCommandLine command = createFlutterSdkRunCommand(workDir, device, mode, path);
+    final GeneralCommandLine command = createFlutterSdkRunCommand(workDir, programArgs, device, mode, path);
```
Comment (skybrian) [inline src/io/flutter/run/daemon/FlutterAppService.java]: Suppose we move createFlutterSdkRunCommand() to SdkFields and just pass in the GeneralCommandLine as a parameter here? That we we build the arguments in one place.

(I was thinking about doing the same for Bazel, as well.)

File: src/io/flutter/run/FlutterConfigurationEditorForm.form [hunk 6]
```diff
@@ -46,6 +46,21 @@
         </constraints>
         <properties/>
       </component>
+      <component id="7656f" class="javax.swing.JLabel">
+        <constraints>
+          <grid row="2" column="0" row-span="1" col-span="1" vsize-policy="0" hsize-policy="0" anchor="8" fill="0" indent="0" use-parent-layout="false"/>
+        </constraints>
+        <properties>
+          <labelFor value="e7876"/>
+          <text value="Program &amp;arguments:"/>
```
Comment (skybrian) [inline src/io/flutter/run/FlutterConfigurationEditorForm.form]: Do we want to call this "Additional arguments" for consistency with Bazel?

File: testSrc/unit/io/flutter/run/SdkFieldsTest.java [hunk 7]
```diff
@@ -21,18 +21,22 @@
   public void shouldReadFieldsFromXml() {
     final Element elt = new Element("test");
     addOption(elt, "filePath", "lib/main.dart");
+    addOption(elt, "programArgs", "--debug");
```
Comment (skybrian) [inline testSrc/unit/io/flutter/run/SdkFieldsTest.java]: What's a more realistic example? (I think --debug" doesn't do anything since it's the default?)

Comment (skybrian) [inline testSrc/unit/io/flutter/run/SdkFieldsTest.java]: Do we need to make sure the user doesn't set flags that are important for debugging? (For example, passing in --no-hot will interfere with hot reloading.)

File: src/io/flutter/run/SdkFields.java (hunk 3 above)
Comment (pq) [inline src/io/flutter/run/SdkFields.java]: This is generated by the formatter.  As a general rule we trust it or change it's settings.

File: src/io/flutter/run/SdkFields.java (hunk 2 above)
Comment (pq) [inline src/io/flutter/run/SdkFields.java]: Same here.

File: src/io/flutter/run/SdkFields.java (hunk 4 above)
Comment (pq) [inline src/io/flutter/run/SdkFields.java]: No objection to updating the formatter.

> This is Google style so I keep forgetting to do it the Jetbrains way.

Yeah and you shouldn't have to.  You should always just format and move on.

File: src/io/flutter/run/FlutterConfigurationEditorForm.form (hunk 6 above)
Comment (pq) [inline src/io/flutter/run/FlutterConfigurationEditorForm.form]: Sure!

I'll update.

File: testSrc/unit/io/flutter/run/SdkFieldsTest.java (hunk 7 above)
Comment (pq) [inline testSrc/unit/io/flutter/run/SdkFieldsTest.java]: Good question.  In general, I'd like to add some validation but haven't thought to hard about it.  I'll add a TODO.

File: src/io/flutter/run/SdkFields.java (hunk 3 above)
Comment (skybrian) [inline src/io/flutter/run/SdkFields.java]: I'm guessing the formatter wants `<p>` to be on a separate line, but it's incorrectly not deleting the `<p>` that's already there. So, if you delete the second `<p>` and reformat, I suspect it won't add it back again?

Comment (pq) [inline src/io/flutter/run/SdkFields.java]: (removing but let's fix the formatter)

File: src/io/flutter/run/SdkFields.java (hunk 2 above)
Comment (pq) [inline src/io/flutter/run/SdkFields.java]: (removing but let's fix the formatter)

File: src/io/flutter/run/daemon/FlutterAppService.java (hunk 5 above)
Comment (pq) [inline src/io/flutter/run/daemon/FlutterAppService.java]: I won't have the time to do this just now; maybe you or I could follow up with another quick change?  (Need to context switch shortly 😞 )

**Comment by pq** [comment]
Updated to label as "Additional args".  PTAL.

**Comment by skybrian** [comment]
LGTM

**Comment by pq** [comment]
Thanks!
